src/core/Carousel/Carousel.tsx

```diff
@@ -0,0 +1,335 @@
+    }
+
+    if (event.key === 'ArrowLeft' || event.key === 'ArrowRight') {
+      setKeysPressed((old) => ({ ...old, [event.key]: true }));
+    }
+
+    switch (event.key) {
+      case 'ArrowLeft': {
+        setCurrentIndex((old) => (slideCount + old - 1) % slideCount);
+        break;
+      }
+      case 'ArrowRight': {
+        setCurrentIndex((old) => (slideCount + old + 1) % slideCount);
+        break;
+      }
+    }
+  };
+
+  const handleKeyUp = (event: React.KeyboardEvent) => {
+    if (event.key === 'ArrowLeft' || event.key === 'ArrowRight') {
+      setKeysPressed((old) => ({ ...old, [event.key]: false }));
+    }
+  };
+
+  return (
+    <section
+      className='iui-carousel'
+      aria-roledescription='carousel'
+      tabIndex={0}
+      onKeyDown={handleKeyDown}
+      onKeyUp={handleKeyUp}
+      {...rest}
+      id={id}
+    >
+      <CarouselContext.Provider
+        value={{
+          currentIndex,
+          setCurrentIndex,
+          slideCount,
+          setSlideCount,
+          keysPressed,
+          idPrefix: id,
+        }}
+      >
+        {children}
+      </CarouselContext.Provider>
+    </section>
+  );
+};
+
+const CarouselSlider = (props: React.ComponentPropsWithoutRef<'ol'>) => {
+  const context = React.useContext(CarouselContext);
+
+  if (!context) {
+    throw new Error('CarouselSlider must be used within Carousel');
+  }
+
+  const { currentIndex, setSlideCount, idPrefix } = context;
+
+  React.useLayoutEffect(() => {
+    setSlideCount(React.Children.count(props.children));
+  }, [props.children, setSlideCount]);
+
+  const items = React.useMemo(
+    () =>
+      React.Children.map(props.children, (child, index) =>
+        React.isValidElement(child)
+          ? React.cloneElement(child, { id: `${idPrefix}--slide-${index}` })
+          : child,
+      ) ?? [],
+    [props.children, idPrefix],
+  );
+
+  return (
+    <ol
+      aria-live='polite'
+      {...props}
+      className={cx('iui-carousel-slider', props.className)}
+      style={{
+        transform: `translateX(-${currentIndex * 100}%)`,
+        ...props.style,
+      }}
+    >
+      {items}
+    </ol>
+  );
+};
+
+const CarouselSlide = (props: React.ComponentPropsWithoutRef<'li'>) => {
+  const { className, children, ...rest } = props;
+
+  return (
+    <li
+      className={cx('iui-carousel-slider-item', className)}
+      role='tabpanel'
+      aria-roledescription='slide'
+      {...rest}
+    >
+      {children}
+    </li>
+  );
+};
+
+const CarouselNavigation = (props: React.ComponentPropsWithoutRef<'nav'>) => {
+  const { className, children, ...rest } = props;
+
+  const context = React.useContext(CarouselContext);
+  if (!context) {
+    throw new Error('CarouselNavigation should be used inside Carousel');
+  }
+
+  const { slideCount, setCurrentIndex, keysPressed } = context;
+
+  return (
+    <nav className={cx('iui-carousel-navigation', className)} {...rest}>
+      <div className='iui-carousel-navigation-left'>
```

Comment:
If you gonna add animation, hover part was interestin as this whole article too
https://www.smashingmagazine.com/2016/07/ten-requirements-for-making-home-page-carousels-work-for-end-users/#the-10-carousel-requirements

Reply:
But..
https://web.dev/carousel-best-practices/#don't-use-autoplay
:D

Reply:
yeah, after spending some time on it, i removed autoplay because of all the requirements (hover, focus, reduced motion, page visibility)